### ISSUE TITLE
`build-std` doesn't work

### ISSUE THREAD

**Issue by aaronmondal** [bazelbuild/rules_rust#2068]
We want to use the `-Zbuild-std` cargo flag to rebuild the standard library sanitized. This practice is requiredy by various sanitizers to work properly, for instance ThreadSanitizer and MemorySanitizer.

Since `build-std` is a Cargo flag it cannot be passed via `extra_rustc_flags`. So I tried adding a Cargo config file like

```toml
[unstable]
build-std = ["core", "std", "alloc", "proc_macro"]
```

to the `cargo_config` in `crates_repository`. That doesn't seem to have any effect at all though. Bug?

AFAICS to get behavior like with `build-std` we'd have to use a override the default BUILD file for the standard library, but that doesn't seem like a nice solution considering that this should work with a flag in the cargo config file.

Maybe I'm overlooking something though and this is already supported somehow?

Original issue: https://github.com/allada/turbo-cache/issues/188#issuecomment-1637103453

**Comment by aaronmondal** [comment]
cc @allada

**Comment by scentini** [comment]
I wouldn't call it a bug, but rather an unsupported scenario. We don't support building the standard library from source out of the box, although it's possible to define a toolchain that does this. @daivinhtran is your proof of concept of defining a toolchain that builds the standard library from sources publicly visible?

**Comment by daivinhtran** [comment]
Inspired by https://rustc-dev-guide.rust-lang.org/building/bootstrapping.html, we can define `rust_toolchain` to build `libstd` from sources. However, such toolchain still needs the existence of prebuilt standard libraries; using prebuilt stdlibs to build `libstd` from sources.

To do so, we need a few components:
1. [`rust_stdlib_filegroup` target](https://android.googlesource.com/platform/prebuilts/rust/+/refs/changes/79/2651079/4/linux-x86/1.69.0/BUILD.bazel#14) for the [prebuilt stdlibs](https://android.googlesource.com/platform/prebuilts/rust/+/refs/changes/79/2651079/4/linux-x86/1.69.0/lib/rustlib/x86_64-unknown-linux-gnu/lib). The prebuilt stdlibs are built by [rust/x.py](https://github.com/rust-lang/rust/tree/master)
2. [base toolchain](https://android.googlesource.com/platform/prebuilts/rust/+/refs/changes/79/2651079/4/toolchain/bootstrap/BUILD.bazel#20) using prebuilt stdlibs from (1)
3. [`rust_library` targets](https://android-review.googlesource.com/c/platform/prebuilts/rust/+/2651079/4/linux-x86/1.69.0/BUILD.bazel#74) for sources
4. [`rust_stdlib_filegroup` target](https://android-review.googlesource.com/c/platform/prebuilts/rust/+/2651079/4/toolchain/bootstrap/BUILD.bazel#45) to group all `rust_library`s from (3)
5.  [final toolchain](https://android-review.googlesource.com/c/platform/prebuilts/rust/+/2651079/4/toolchain/bootstrap/BUILD.bazel#62) using stdlibs from (4). This toolchain is registered from WORKSPACE.

Looking at these components, there's a dependency cycle 5 -> 4 -> 3 -> 5 because (3) needs a rust_toolchain to build itself. To break the dependency cycle, we implement [`with_base_transtion` rule](https://android-review.googlesource.com/c/platform/prebuilts/rust/+/2651079) to wrap the files from (4) and attach an incoming transition to set `extra_toolchains` to the base toolchain in (2). During toolchain resolution, base toolchain set by `extra_toolchains` command line option is preferred/selected to compile the stdlibs's sources - breaking the dependency cycle into 5 -> 4 -> 3 -> 2 -> 1.

@aran @aaronmondal let me know if you have any issue setting this up. I think building the prebuilt stdlibs (1) correctly is the main heavy-lifting to get this working.

**Comment by daivinhtran** [comment]
@aaronmondal @aran 

This thread has been a while. I finally got some time for this.

https://github.com/daivinhtran/rules_rust/tree/toolchain-to-rebuild-std/examples/toolchain-to-rebuild-std is an example how to bootstrap a rust toolchain and rebuild libstd. If you're using a target triple that's already in https://github.com/bazelbuild/rules_rust/blob/main/rust/platform/triple_mappings.bzl, it should be easy to follow the example.

https://github.com/daivinhtran/rules_rust/blob/toolchain-to-rebuild-std/examples/toolchain-to-rebuild-std/stdlibs.BUILD.bazel is the BUILD file for defining the standard library. You can add any custom configuration to the `rust_library` targets.

**Comment by lylythechosenone** [comment]
I made that issue to supersede this one, since it seems to be dead and was originally based on a flawed concept, but (maintainers) please feel free to choose which you'd like to follow. Regardless, this is a very important topic, since it is required in order for `rules_rust` to support bare-metal and custom targets.

**Comment by esqu1** [comment]
@lylythechosenone -- any progress on this issue at all? This currently blocks building C++ with threading support that also makes use of Rust libraries:

```
error: --shared-memory is disallowed by std-1d3ccab4f7990fa9.std.46da483ec1cddbbe-cgu.0.rcgu.o because it was not compiled with 'atomics' or 'bulk-memory' features.
```

I did see your proof-of-concept commit and was wondering how to utilize the `rust_src` target downstream?

**Comment by lylythechosenone** [comment]
Sadly, I haven't gotten very far on this at all. That proof-of-concept is all I've got at this point.

**Comment by blaizard** [comment]
I wanted to share a slightly different method I am using: instead of having 2 toolchains (1 to bootstrap and the final one), I use the same (only 1) with a `select` branch on the `rust_std` attribute toward the prebuilt libs (1) and the other branch toward the `rust_library` (3) to build the new std lib from source.
On the `rust_library` I apply a transition (`with_base_transtion` rule) to flip a config setting to tell to use the prebuilt libs branch from the `rust_std` attribute.
So in short with this method, I replace (2) from @tidefield method with a `select` + config setting.

**Comment by rejuvenile** [comment]
I had Claude modify rules_rust to build-std, but the change is somewhat specific to my toolchain. It is possible.
